CORE-4190: Support Python3 & fix pure python tests

chpedersen opened the pull request:
* Most changes are trivial:
  - `"foo"` -> `b"foo"`, `bytes/str/unicode`
  - `iteritems`/`itervalues`/`range`/`xrange` etc.
* In python3 there's no `buffer` so I used `memoryview`. The buffer constructor takes offset & size args, but memoryview doesn't, so I use the `buffer_with_offset` func.
* Some blocks are repeated several times:
```
if six.PY3:
    dtype = obj_dtype.char.encode()
    if cython.compiled:
        cdtype = dtype[0]
    else:
        cdtype = bytes([dtype[0]])
else:
    dtype = obj_dtype.char
    cdtype = cython.cast('const char*', dtype)[0]
```
maybe it's just more efficient to keep them rather than move this to a separate function, IDK
* I don't remeber exactly how, but dict comparison is no the same in py3 as in py2, so I stole the `dict_cmp` from https://stackoverflow.com/questions/3484293/is-there-a-description-of-how-cmp-works-for-dict-objects-in-python-2
* I didn't do big changes in tests, just the "easy" ones. All tests pass :raised_hands: BUT The tests in python3 are slower than in python2, even with cython. In particular, 
`MappedStringMappingBigTest` took about 5 minutes vs 1.5 seconds in python2 :open_mouth: 

Sorry for the not very descriptive 100+ commit messages

chpedersen committed:
CORE-4190: iteritems

chpedersen committed:
CORE-4190: xrange

chpedersen committed:
CORE-4190: itervalues

chpedersen committed:
CORE-4190: iterkeys

chpedersen committed:
CORE-4190: zip

chpedersen committed:
CORE-4190: zip_longest

chpedersen committed:
CORE-4190: itervalues

chpedersen committed:
CORE-4190: L suffix

chpedersen committed:
CORE-4190: raise Exception

chpedersen committed:
CORE-4190: cPickle

chpedersen committed:
CORE-4190: clru cache instead of chorde's

chpedersen committed:
CORE-4190: is_cpython

chpedersen committed:
CORE-4190: unicode -> text_type

chpedersen committed:
CORE-4190: long -> int

chpedersen committed:
CORE-4190: unicode -> text_type

chpedersen committed:
CORE-4190: buffer -> memoryview

chpedersen committed:
CORE-4190: unicode -> text_type

chpedersen committed:
CORE-4190: unicode -> text_type

chpedersen committed:
CORE-4190: safe_utf8

chpedersen committed:
CORE-4190: unicode -> text_type

chpedersen committed:
CORE-4190: buffer & long

chpedersen committed:
CORE-4190: iteritems

chpedersen committed:
CORE-4190: list(map())

chpedersen committed:
CORE-4190: list(map())

chpedersen committed:
CORE-4190: int -> str -> bytes

chpedersen committed:
CORE-4190: xrange/range

chpedersen committed:
CORE-4190: int division

chpedersen committed:
CORE-4190: basestring

chpedersen committed:
CORE-4190: assign int to buf

chpedersen committed:
CORE-4190: assign bytes to buf

chpedersen committed:
CORE-4190: assign int to buf

chpedersen committed:
CORE-4190: iterkeys

chpedersen committed:
CORE-4190: assign int to buf

chpedersen committed:
CORE-4190: encode

chpedersen committed:
CORE-4190: read byte from buf

chpedersen committed:
CORE-4190: long/basestring/buffer

chpedersen committed:
CORE-4190: bytes literal

chpedersen committed:
CORE-4190: add missing cmp function

chpedersen committed:
CORE-4190: typecode string

chpedersen committed:
CORE-4190: mapped_list unpack_from

chpedersen committed:
CORE-4190: mapped_frozenset.unpack_from

chpedersen committed:
CORE-4190: proxied_list.metadata

chpedersen committed:
CORE-4190: buffer/memoryview

chpedersen committed:
CORE-4190: write bytes

chpedersen committed:
CORE-4190: str -> bytes

chpedersen committed:
CORE-4190: iterkeys

chpedersen committed:
CORE-4190: buffer_with_offset

chpedersen committed:
CORE-4190: str -> bytes

chpedersen committed:
CORE-4190: mapped_tuple.pack_info int offset

chpedersen committed:
CORE-4190: mapped_list.unpack_from buf.to_bytes

chpedersen committed:
CORE-4190: array dtype.decode

chpedersen committed:
CORE-4190: mapped_tuple.pack_into all_floats size_bytes

chpedersen committed:
CORE-4190: str -> bytes literal

chpedersen committed:
CORE-4190: proxied_list._metadata dcode 'd'

chpedersen committed:
CORE-4190: fix for LongLongIntContainerPackingTest

chpedersen committed:
CORE-4190: fix for BytesContainerPackingTest

chpedersen committed:
CORE-4190: get_wrapped_key

chpedersen committed:
CORE-4190: buffer_with_offset

chpedersen committed:
CORE-4190: BytesContainerPackingTest

chpedersen committed:
CORE-4190: proxied_list cmp operators

chpedersen committed:
CORE-4190: str -> bytes

chpedersen committed:
CORE-4190: mapped_list.unpack_from buf.tobytes if not bytes

chpedersen committed:
CORE-4190: proxied_buffer.unpack_from buffer_with_offset

chpedersen committed:
CORE-4190: str -> bytes

chpedersen committed:
CORE-4190: iteritems

chpedersen committed:
CORE-4190: buffer_with_offset

chpedersen committed:
CORE-4190: buffer_with_offset default size

chpedersen committed:
CORE-4190: list(dictitems)

chpedersen committed:
CORE-4190: fix for NumpyCastingContainerPackingTest

chpedersen committed:
CORE-4190: fix for NumpyCastingContainerPackingTest

chpedersen committed:
CORE-4190: dtype: bytes

chpedersen committed:
CORE-4190: proxied_buffer.pack_into objlen fix

chpedersen committed:
CORE-4190: iteritems

chpedersen committed:
CORE-4190: DictContainerPackingTest

chpedersen committed:
CORE-4190: NestedObjectPackaging test

chpedersen committed:
CORE-4190: MappedStringMappingTest

chpedersen committed:
CORE-4190: iteritems

chpedersen committed:
CORE-4190: DictPackingCommonTest

chpedersen committed:
CORE-4190: str -> bytes

chpedersen committed:
CORE-4190: proxied_list_cmp None

chpedersen committed:
CORE-4190: fix use map result twice -> list(map)

chpedersen committed:
CORE-4190: _map_zipfile compress_size

chpedersen committed:
CORE-4190: map_file fileno

chpedersen committed:
CORE-4190: buffer_with_offset

chpedersen committed:
CORE-4190: fix fileno in map_file

chpedersen committed:
CORE-4190: fix fileno in map_file

chpedersen committed:
CORE-4190: buffer_with_offset

chpedersen committed:
CORE-4190: str -> bytes

chpedersen committed:
CORE-4190: str -> bytes

chpedersen committed:
CORE-4190: str -> bytes

chpedersen committed:
CORE-4190: str -> bytes

chpedersen committed:
CORE-4190: str -> bytes

chpedersen committed:
CORE-4190: str -> bytes

chpedersen committed:
CORE-4190: mapped_frozenset.pack_into npndarray

chpedersen committed:
CORE-4190: fix ApproxIntegerIdMultiMapperTest

chpedersen committed:
CORE-4190: fix stable_hash

chpedersen committed:
CORE-4190: comment

chpedersen committed:
CORE-4190: comment

chpedersen committed:
CORE-4190: add proxied_list.__eq__ and remove __richcmp__

chpedersen committed:
CORE-4190: add proxied_list.__eq__ and remove __richcmp__

chpedersen committed:
CORE-4190: str -> bytes

chpedersen committed:
CORE-4190: str.encode

chpedersen committed:
CORE-4190: cython

chpedersen committed:
CORE-4190: fix bsearch not implemented

chpedersen committed:
CORE-4190: fix objlen -> memoryview.nbytes

chpedersen committed:
CORE-4190: fix proxied list iter & getter

chpedersen committed:
CORE-4190: fix proxied_frozenset tests

chpedersen committed:
CORE-4190: fix proxied_frozenset tests

chpedersen committed:
CORE-4190: fix proxied_frozenset tests

chpedersen committed:
CORE-4190: tests.pool iteritems-xrange

chpedersen committed:
CORE-4190: buffer

chpedersen committed:
CORE-4190: str -> bytes

chpedersen committed:
CORE-4190: fix buffer py3

chpedersen committed:
CORE-4190: fix buffer py3

chpedersen committed:
CORE-4190: fix proxied_frozenset subet

klaussfreire commented:
>     * I didn't do big changes in tests, just the "easy" ones. All tests pass raised_hands BUT The tests in python3 are slower than in python2, even with cython. In particular,
>       `MappedStringMappingBigTest` took about 5 minutes vs 1.5 seconds in python2 open_mouth

That's not a small thing. That's a sign of a serious problem somewhere.

klaussfreire commented on tests/mapped_struct.py:
```diff
@@ -855,11 +862,11 @@ def gen_values(self, n, reversed = False, shuffled = False, gen_dupes = False):
             return itertools.chain(
-                itertools.izip(keys, xrange(0, 2*n, 2)),
-                itertools.islice(itertools.izip(keys, xrange(0, 2*n, 2)), 10, None),
+                zip(keys, xrange(0, 2*n, 2)),
+                itertools.islice(zip(keys, xrange(0, 2*n, 2)), 10, None),
```
This materializes the sequences in py2 and will use a lot of RAM, use izip in py2:

``` python
from future.builtins import zip
```

That requires future, but it's a good way to get conisstent behavior.

There's also `map` BTW.

chpedersen replied:
There's the `from six.moves import xxx` for this

klaussfreire replied:
Hadn't seen it

klaussfreire commented on tests/mapped_struct.py:
```diff
@@ -1007,9 +1014,9 @@ def _testBuild(self, N, tempdir, **gen_kwargs):
             if isinstance(elem, list):
                 for e in elem:
-                    self.assertTrue(isinstance(e, (int, long)))
+                    self.assertTrue(isinstance(e, (int, long, numpy.uint32)))
```
No, we can't allow numpy types here.

If this is for pure python tests, I'd say rollback. We can require Cython.

olifer97 replied:
why not numpy types?

klaussfreire replied:
The original input were native types. If the library produces numpy types on output, something is broken.

klaussfreire commented on tests/mapped_struct.py:
```diff
@@ -1094,7 +1101,7 @@ def testBuildValueDedup(self):
         if isinstance(test_values, dict):
             test_values = test_values.items()
```
```suggestion
            test_values = iteritems(test_values)
```

klaussfreire commented on tests/mapped_struct.py:
```diff
@@ -1292,17 +1299,17 @@ class MappedApproxString32MultiMappingTest(MappedApproxStringMultiMappingTest):
 
 class MappedStringMappingUnicodeTest(MappedStringMappingTest):
-    TEST_KEYS = [ "%s€ ···YEAH···" % (k,) for k in MappedStringMappingTest.TEST_KEYS ]
+    TEST_KEYS = [b'%s\xe2\x82\xac \xc2\xb7\xc2\xb7\xc2\xb7YEAH\xc2\xb7\xc2\xb7\xc2\xb7' % (k,) for k in MappedStringMappingTest.TEST_KEYS ]
```
Unreadable.

Use `u"%s€ ···YEAH···".encode("utf8")` maybe?

klaussfreire commented on tests/mapped_struct.py:
```diff
@@ -1314,10 +1321,10 @@ class BsearchTest(unittest.TestCase):
             random.shuffle(testarray)
             a = numpy.array(testarray[:50], dtype)
-            b = numpy.array([0] + testarray[50:] + range(101,103), dtype)
+            b = numpy.array([0] + testarray[50:] + list(range(101,103)), dtype)
```
``` python
from past.builtins import range, xrange
```

And then you don't have to change these and it works as expected in all cases.

FTR, past and future helpers are only good for tests and pure python code. Cythonized code can't benefit from these because the wrappers aren't recognized in optimizations, so it generates horrible code for their use.

klaussfreire commented on tests/mapped_struct.py:
```diff
@@ -1731,7 +1771,7 @@ def testMixed(self):
         self.assertFalse(big > big)
 
-        big = self.pack([1L, 2L, 3L])
+        big = self.pack([1, 2, 3])
```
They need to be longs. Use `long(1)` maybe as an easy way to make it dual python?

klaussfreire commented on tests/mapped_struct.py:
```diff
@@ -2082,10 +2122,24 @@ def testProxiedTupleRepr(self):
 class DictPackingCommonTest(object):
 
     def assertUnsortedEquals(self, a, b):
+        if six.PY3:
+            return sorted(a, key=functools.cmp_to_key(key_cmp)) == sorted(b, key=functools.cmp_to_key(key_cmp))
```
I don't follow why you need this. Sorting item tuples should work fine in both python versions.

chpedersen replied:
No, it's comparing int and str which doesn't work in py3. For example, from one of the values in MappedDictPackingTest:

in py2
```
sorted(['a', 1, frozenset(), (1, 2)]) == [1, frozenset([]), 'a', (1, 2)]
```
in py3
```
>>> sorted(['a', 1, frozenset(), (1, 2)])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'int' and 'str'
```

klaussfreire replied:
Oh... it's a mixed-type test, didn't notice.

What about `key=repr`?

klaussfreire commented on tests/mapped_struct.py:
```diff
@@ -2187,11 +2241,11 @@ def testHashIntegersFixed(self):
             (66210231110, 66210231110),
             (752, 752),
-            (-1, 18446744073709551615L),
+            (-1, 18446744073709551615),
```
Again, use the long(x) idiom

klaussfreire reviewed (changes requested):
Partial review only of tests (seemed simpler to start with)

chpedersen committed:
CORE-4190: long

chpedersen committed:
CORE-4190: iteritems

chpedersen committed:
CORE-4190: unicode encode

chpedersen committed:
CORE-4190: memoryview bytes (itemsize=1)

chpedersen committed:
CORE-4190: dont multiply len x itemsize

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -438,7 +438,7 @@ def shared_id(obj):
         elif isinstance(obj, long):
             # Real longs must make room for the flag bits
-            return (obj & 0xFFFFFFFFFFFFFFFF) | ((obj >> 64) << 68) | LONG_MASK
+            return (obj & long(0xFFFFFFFFFFFFFFFF)) | ((obj >> 64) << 68) | LONG_MASK
```
This will probably create a python int, not native code. Have you checked the output of `cython --annoate`?

klaussfreire replied:
BTW... the `long(x)` idiom is only for tests because of this (inefficient). It's ok in tests, we don't care, but actual library code should be efficient.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -742,15 +742,15 @@ def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0,
                         buf[offs] = ord(dtype)
-                    elif (cython.cast(cython.longlong, -0x8000000000000000) <= iminval
-                            and imaxval <= cython.cast(cython.longlong, 0x7FFFFFFFFFFFFFFF)):
+                    elif (cython.cast(cython.longlong, long(-0x8000000000000000)) <= iminval
+                            and imaxval <= cython.cast(cython.longlong, long(0x7FFFFFFFFFFFFFFF))):
```
Idem. Why this change? It should not have been affected by py3, and this way won't be optimal code IIUC.

chpedersen replied:
Ok, I see that calling `long` is expensive here. But why did we have the "L" suffix in the first place?  It seems to me it doesn't change anything

```python
cython.cast(cython.longlong, -0x8000000000000000L)
```

turns into
```c
(PY_LONG_LONG)-9223372036854775808L)
```

and 
```python
cython.cast(cython.longlong, -0x8000000000000000)
```
into
```c
(PY_LONG_LONG)-9223372036854775808LL)
```

klaussfreire replied:
Did you try python 2?

klaussfreire replied:
TBH, I think the correct version is the second version, because it produces a LL literal, which is the right kind. But I suspect that's not the case in py2 because we've had issues with that in the past, where it would use a regular int literal that gets truncated.

Truncation produces a warning, so you can check for warnings while building.

klaussfreire replied:
py2:
```
+4: x = cython.cast(cython.longlong, -0x8000000000000000)

  __pyx_v_4test_x = ((PY_LONG_LONG)-9223372036854775808LL);
```

py3
```
+4: x = cython.cast(cython.longlong, -0x8000000000000000)

  __pyx_v_4test_x = ((PY_LONG_LONG)-9223372036854775808LL);
```

klaussfreire replied:
Looks like removing the L suffix is the right way to go about it. Must have been fixed in sme cython patch because it was needed before.

`\_(0_0)_/`

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -125,13 +125,16 @@
     basestring = (bytes, str)
 
+def make_memoryview(a):
+    return memoryview(a).cast("B")
```
Why don't you make it a cfunc, inline, and put hte `if six.PY2` check inside?

Make sure to cache the `six.PY2` bit in a native global to avoid slow python access inside it.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -141,7 +144,7 @@
 else:
     if six.PY3:
-        buffer = memoryview
+        buffer = make_memoryview
```
In py2, memoryview has no `cast`, so this doesn't work as-is.

chpedersen replied:
`make_memoryview` is only used in py3

klaussfreire replied:
You can't do that with make_memoryview as a cfunc. You have to make make_memoryview work in both versions, and have all the python2 conditionals inside it.

That's because you can't assign a cfunc to a python variable. I'm not sure how it's working for you now in fact. It should break hard.

klaussfreire replied:
ping

klaussfreire replied:
You may be able to avoid the `buffer` global and replace all uses of buffer with make_memoryview if you do that BTW. It will work better.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -967,7 +967,10 @@ def unpack_from(cls, buf, offs, idmap = None, array = array.array, klass = list)
-            rv = array(dtype, buf[offs:offs+itemsize*objlen])
+            w = buf[offs:offs+itemsize*objlen]
+            if six.PY3:
+                w = bytes(w)
```
This will copy

olifer97 commented on tests/mapped_struct.py:
```diff
@@ -12,20 +12,27 @@
 from datetime import datetime, date
 from decimal import Decimal
 import numpy as np
+import six
+
+if six.PY3:
+    long = int
+    buffer = memoryview
```
this is because we will be using this library in python2 and python3 right?

olifer97 commented on tests/mapped_struct.py:
```diff
@@ -1699,6 +1706,39 @@ def testSubsetSuperset(self):
         self.assertTrue(small < big)
         self.assertTrue(big > small)
 
+        a = self.pack([1., 2.])
+        b = self.pack([1., 2., 3., 4.])
+        self.assertTrue(a < b)
+        self.assertTrue(a <= b)
+        self.assertFalse(b < a)
+        self.assertFalse(b <= a)
+        self.assertNotEquals(a, b)
+
+        a = self.pack([1., 2., 3.])
+        b = self.pack([1., 2.5, 3., 4.])
+        self.assertFalse(a < b)
+        self.assertFalse(a <= b)
+        self.assertFalse(b < a)
+        self.assertFalse(b <= a)
+        self.assertNotEquals(a, b)
+
+        a = self.pack([1., 2.5, 3.])
+        b = self.pack([1., 2., 3., 4.])
+        self.assertFalse(a < b)
+        self.assertFalse(a <= b)
+        self.assertFalse(b < a)
+        self.assertFalse(b <= a)
+        self.assertNotEquals(a, b)
+
+        a = self.pack([1., 2., 3., 10.])
+        b = self.pack([1., 2., 3., 4.])
+        self.assertFalse(a < b)
+        self.assertFalse(a <= b)
+        self.assertFalse(b < a)
+        self.assertFalse(b <= a)
+        self.assertNotEquals(a, b)
```
why do we have to add all this asserts after migrating to python3? the behavior shouldn't be the same?

olifer97 reviewed (commented):
The idea is to have this library support both python2 and python3 because we dont have all the components that used this library migrated right?
I dont think i have enough knowledge to approved this, the problem with the duration of tests seems like a severe problem if we use this library in places perfomance-sensitive.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -462,30 +492,43 @@ def is_wrapped_key(obj):
-                dtype = obj_dtype.char
-                cdtype = cython.cast('const char*', dtype)[0]
-                if cdtype in ('l', 'I', 'i', 'H', 'h', 'B', 'b'):
+                if six.PY3:
+                    dtype = obj_dtype.char.encode()
+                    if cython.compiled:
+                        cdtype = dtype[0]
+                    else:
+                        cdtype = bytes([dtype[0]])
+                else:
+                    dtype = obj_dtype.char
+                    cdtype = cython.cast('const char*', dtype)[0]
```
You can't have this check (six.py3) in performance-critical code, as it compiles to a python attribute access on the global module. A relatively slow operation.

Store the result of this check into a cython.bint global variable, and use that variable instead if you need to.

Otherwise, try to write version-agnostic code.

In this case, if you declare dtype as unicode, `dtype[0] in ('l', 'I', 'i', ...)` generates pretty good code if you take a look. You would need 2 verisons of the code, py2 and py3, but the py3 version would actually be quite simpler (no char pointer stuff).

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -575,22 +618,28 @@ def unpack_from(cls, buf, offs, idmap = None):
         try:
-            fs_type = pbuf[offs]
-            if fs_type == 'm' or fs_type == 'M':
+            if six.PY3 and not cython.compiled:
```
cython.compiled is a static check, so always do it first, as it's the fastest

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -81,13 +82,19 @@
 from datetime import timedelta, datetime, date
 from decimal import Decimal
+from six import itervalues, reraise, iteritems, iterkeys
+from six.moves import cPickle, range
```
importing range may make cython's loop optimizations go bad.

Check cython --annotate

chpedersen replied:
oops, yes. i'll look into it

chpedersen replied:
We could just use `range` without importing, although in py2 without cython it would be slower than xrange of course (py2 + cython is ok with `range`). I don't see another way around this.

klaussfreire replied:
You can trust cython to properly implement xrange, I think.

Keep unapologetically using xrange, cython will generate proper code for it even if it's not there in py3.

Cython is not optional after all, lets remember that. The fact that tests are failing in pure python is a sign of that.

If you want to cater to pure python, you can do the globals trick:

``` python
if not cython.compiled:
     from six.moves import range as _range
     globals().update(dict(range=_range))
```

That has the same effect as an import in pure python, but is invisible to Cython.

klaussfreire replied:
ping. This is breaking cython optimizations. Really needs fixing.

chpedersen committed:
CORE-4190: remove unnecessary call to long

chpedersen committed:
CORE-4190: global py2/3 flags

chpedersen committed:
CORE-4190: long -> cython cast

chpedersen committed:
CORE-4190: make_memoryview cfunc/inline

chpedersen committed:
CORE-4190: cast instead of copy

chpedersen committed:
CORE-4190: cython.compiled first in if condition

chpedersen committed:
CORE-4190: dtype no encode

chpedersen committed:
CORE-4190: ord constant instead of dtype

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -441,7 +446,7 @@ def shared_id(obj):
         elif isinstance(obj, long):
             # Real longs must make room for the flag bits
-            return (obj & long(0xFFFFFFFFFFFFFFFF)) | ((obj >> 64) << 68) | LONG_MASK
+            return (obj & cython.cast(cython.longlong, 0xFFFFFFFFFFFFFFFF)) | ((obj >> 64) << 68) | LONG_MASK
```
longlong can't handle that literal. It needs an ulonglong, but since you're doing `obj &` you really need a python long.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -745,15 +750,15 @@ def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0,
                         raise OverflowError
                 except OverflowError:
-                    if 0 <= minval and maxval <= long(0xFFFFFFFFFFFFFFFF):
+                    if 0 <= minval and maxval <= cython.cast(cython.longlong, 0xFFFFFFFFFFFFFFFF):
```
longlong can't handle that literal. You need an ulonglong, and IIRC using an ulonglong generates ugly code, so a python long is a better choice. I think using simply `0xFFFFFFFFFFFFFFFF` should work in both versions, but do verify with the annotated code.

klaussfreire replied:
ping

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -103,6 +103,9 @@
 import cython
 
+python2 = cython.declare(cython.bint, six.PY2)
+python3 = cython.declare(cython.bint, six.PY3)
```
I know it sounds unnecessary, but color me paranoid:

``` python
assert python2 or python3
```

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -1377,7 +1382,7 @@ def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0):
         cur_offs = offs
-        if six.PY3:
+        if python3:
             objlen = obj.nbytes
```
I'm not sure about using .nbytes just like that.

Why don't you take a memoryview explicitly to do the cast to bytes?

We want to be polymorphic here and accept byte strings, bytearrays, and iterables of int. Using nbytes straightaway breaks for byte strings and bytearrays both.

In fact, maybe add a test to make sure this works with byte stirngs and byte arrays.

chpedersen replied:
> I'm not sure about using .nbytes just like that.

ok, I changed this

> We want to be polymorphic here and accept byte strings, bytearrays, and iterables of int. Using nbytes straightaway breaks for byte strings and bytearrays both.

`proxied_buffer.unpack_from` returns a `buffer`. So the `self.assertEquals(obj, unpacked_obj)` fails if `obj` is a string (bytes) in py2. (in py3 it works, `memoryview(obj) == obj` is true)

klaussfreire replied:
Maybe the test can cast the expected object to buffer in the assert then:

``` python
self.assertEquals(buffer(obj), unpacked_obj)
```

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -950,9 +955,9 @@ def unpack_from(cls, buf, offs, idmap = None, array = array.array, klass = list)
-                w = bytes(w)
+            w = buf[offs:offs + itemsize * objlen]
+            if python3:
+                w = memoryview(w).cast(dtype)
```
Isn't it a memoryview already by now?

It comes from likerobuffer.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -965,7 +970,7 @@ def unpack_from(cls, buf, offs, idmap = None, array = array.array, klass = list)
             q = buf[offs:offs + itemsize * objlen]
-            if six.PY3:
+            if python3:
                 q = bytes(q)
```
All those bytes(q) will also copy. Those also need fixing.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -143,16 +148,16 @@ def make_memoryview(a):
-        if six.PY3:
+        if python3:
             return buffer(data)[offset:offset+size]
         return buffer(data, offset, size)
```
I feel more fuzzy with an else in there

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -492,7 +497,7 @@ def is_wrapped_key(obj):
+    if python2:
         if isinstance(obj, tuple):
             return obj[1]
         elif isinstance(obj, int):
```
Again, I prefer else for exclusive paths:

``` python
if python2:
   # py2 verison
else:
   # py3 verison
```

Instead of

``` python
if python3:
    # py3 version
# py2 version
```

It's just so much more readable in the first form.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -514,27 +514,29 @@ def get_wrapped_key(obj):
 
     @classmethod
-    @cython.locals(cbuf = 'unsigned char[:]', i=int, ix=int, offs=cython.longlong, cdtype=cython.uchar)
+    @cython.locals(cbuf = 'unsigned char[:]', i=int, ix=int, offs=cython.longlong, py2_dtype=cython.uchar, py3_dtype="unicode")
```
Instead of py2/3_dtype, cdtype and udtype?

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -514,27 +514,29 @@ def get_wrapped_key(obj):
-                    all_intlong = 1
-                elif cdtype in (b'd', b'f'):
-                    all_float = 1
+                    temp = obj_dtype.char
+                    py2_dtype = cython.cast('const char*', temp)[0]
```
Add a comment stating why you need temp (to keep a reference to the string while you access the char*)

Otherwise I can totally see someone thinking it's not necessary and causing issues later on

chpedersen committed:
CORE-4190: rename -> cdtype/udtype

chpedersen committed:
CORE-4190: if/else

chpedersen committed:
CORE-4190: fix cast

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -504,38 +506,38 @@ def get_wrapped_key(obj):
+        if isinstance(obj, tuple):
+            return obj[1]
+        elif isinstance(obj, int):
+            return obj & ~WRAP_MASK
```
You can simplify this.

The python2 and 3 sides are identical except for the `long` case.

``` python
if isinstance(obj, tuple):
    return obj[1]
elif python2:
    # py2 cases
else:
    # py3 cases
```

That way you avoid repeating the common case.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -575,22 +627,28 @@ def unpack_from(cls, buf, offs, idmap = None):
         else:
             pbuf = buf
         try:
-            fs_type = pbuf[offs]
-            if fs_type == 'm' or fs_type == 'M':
+            if not cython.compiled and python3:
+                fs_type = bytes([pbuf[offs]])
+            else:
+                fs_type = pbuf[offs]
```
This is unnecessary. `pbuf` is an `uchar*` which means `pbuf[offs]` will be an `uchar` regardless of python version.

If this is catering to pure python runs, remember we don't need pure python support. We should drop it and avoid the complications it brings, which means you should revert these to the cython-native implementation.

We should probably switch all `if cython.compiled` into `assert cython.compiled` (in another PR) because we should really drop pure python support. It's way too hard to maintain and completely unnecessary.

Pure python support was meant to allow running natively in PyPy, but PyPy's cpyext improved to the point where it can run cythonized code quite robustly, so we can depend on cython and simplify all those pesky `if cython.compiled` cases.

So... for now, just forget about pure python tests. Worry about making it run in cython.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -716,30 +788,30 @@ def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0,
             buf[offs:offs+len(abuf)] = abuf
             offs += len(abuf)
-            offs = (offs + 7) / 8 * 8
+            offs = (offs + 7) // 8 * 8
```
The best way to avoid headaches with division semantics in dual code, is to add `from __future__ import division` to the imports on the top, and make sure we always consistently use py3 semantics.

This means using `//` here as you did, but without the future import this will break py2.

klaussfreire replied:
Ping. Add the future import.

chpedersen replied:
It's already there :P

klaussfreire replied:
No... where?

klaussfreire replied:
NVM... I was clicking the file name and github was sending me to an old version of the file. I see it now.

chpedersen committed:
CORE-4190: assert python version

chpedersen committed:
CORE-4190: .cast instead of bytes()

chpedersen committed:
CORE-4190: simplify get_wrapped_key

chpedersen committed:
CORE-4190: import division

chpedersen committed:
CORE-4190: revert bytes pbuf

chpedersen committed:
CORE-4190: dtype no encode

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -501,17 +503,15 @@ def is_wrapped_key(obj):
     # the key for the unwrapped value
+    if isinstance(obj, tuple):
+        return obj[1]
     if python2:
```
```suggestion
    elif python2:
```

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -776,7 +788,12 @@ def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0,
+            if check_udtype:
+                l_or_L = udtype[0] == u'l' or udtype[0] == u'L'
+            else:
+                l_or_L = dtype == b'l' or dtype == b'L'
```
There are cases in py3 where you can end up here. Wouldn't that fail the comparison because of unicode-bytes mismatch?

Maybe always set udtype and instead of "check_udtype" just use "if python3"

chpedersen committed:
Update sharedbuffers/mapped_struct.py

Co-authored-by: klaussfreire <klaussfreire@gmail.com>

chpedersen committed:
CORE-4190: use udtype py3

chpedersen committed:
CORE-4190: add comment

chpedersen committed:
CORE-4190: use len(obj)

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -671,18 +671,18 @@ class mapped_tuple(tuple):
         objlen = len(obj)
-        check_udtype = False
+        udtype = u''
+        dtype = b''
```
These are not valid dtypes. I'd rather use None.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -807,7 +817,10 @@ def pack_into(cls, obj, buf, offs, idmap = None, implicit_offs = 0,
+                if python3:
+                    a = array(udtype, obj)
+                else:
+                    a = array(dtype, obj)
```
```suggestion
                a = array(udtype if python3 else dtype, obj)
```

chpedersen committed:
Update sharedbuffers/mapped_struct.py

Co-authored-by: klaussfreire <klaussfreire@gmail.com>

chpedersen committed:
CORE-4190: remove cast to longlong

chpedersen committed:
CORE-4190: init dtypes to None

chpedersen committed:
CORE-4190: remove cast to longlong

chpedersen committed:
CORE-4190: test proxied buffer packing

chpedersen committed:
CORE-4190: test proxied buffer packing cast to buffer

klaussfreire commented:
I'm getting a failure in python 3.9:

```
======================================================================
FAIL: testProxiedTupleHash (tests.mapped_struct.ProxiedTuplePackingTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/claudiofreire/src/sharedbuffers/tests/mapped_struct.py", line 2075, in testProxiedTupleHash
    self.assertTrue(hash(p) == hash(t))
AssertionError: False is not true

----------------------------------------------------------------------
Ran 637 tests in 1901.311s
```

I suspect it has something to do with python using randomized hash seeds.

The tests took also significantly longer than in py2, with py2 doing them all in 1498s and py3 1901s. There's probably something slow in the new code, maybe memory being copied that shouldn't or something like that.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -1111,6 +1236,10 @@ def seek(self, pos):
+        return (a > b) - (a < b)
+
 @cython.cclass
 class proxied_dict(object):
```
You need to change proxied_dict to have python 3's dict interface: keys, values, items are all iterators now. In fact keys yields a set-like interface, ideally we'd return a proxied_set kind of thing. It will take some doing. Maybe for a future PR, but for now, yielding iterators is good enough.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -114,17 +128,44 @@
+def buffer_with_offset(data, offset, size=None):
+    if size is not None:
+        if python3:
+            return buffer(data)[offset:offset+size]
+        else:
+            return buffer(data, offset, size)
```
Should call make_memoryview directly (in both python versions), to make use of it being cdef. Have the py3 conditionals inside make_memoryview. Been saying it for a while.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -6326,7 +6595,14 @@ def build(cls, initializer, destfile = None, tempdir = None,
 
             indexpos = curpos
-            write(buffer(index))
+            if python3:
+                if len(index):
+                    w = buffer(index)
+                else:
+                    w = b''
+            else:
+                w = buffer(index)
```
This does not seem necessary. I'm perfectly able to do a memoryview of an empty array in py3. What exactly prompted this?

chpedersen replied:
It breaks when `index = array([], shape=(0, 2), dtype=uint64)`
```
File "sharedbuffers/mapped_struct.py", line 139, in sharedbuffers.mapped_struct.make_memoryview
    return memoryview(a).cast("B")
TypeError: memoryview: cannot cast view with zeros in shape or strides
```

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -464,29 +505,42 @@ def get_wrapped_key(obj):
-                elif cdtype in ('d', 'f'):
-                    all_float = 1
+                if python3:
+                    udtype = obj_dtype.char
+                    if udtype[0] in (u'l', u'I', u'i', u'H', u'h', u'B', u'b'):
+                        all_int = all_intlong = 1
+                    elif udtype[0] == u'L':
+                        all_intlong = 1
+                    elif udtype[0] in (u'd', u'f'):
+                        all_float = 1
```
```suggestion
                    udtype = obj_dtype.char
                    udtype0 = udtype[0]
                    if udtype0 in (u'l', u'I', u'i', u'H', u'h', u'B', u'b'):
                        all_int = all_intlong = 1
                    elif udtype0 == u'L':
                        all_intlong = 1
                    elif udtype0 in (u'd', u'f'):
                        all_float = 1
```

Seems to generate substantially better code

chpedersen committed:
CORE-4190: use make_memoryview instead of buffer

chpedersen committed:
CORE-4190: use buffer_with_offset

chpedersen committed:
CORE-4190: remove cython not compiled case

chpedersen committed:
CORE-4190: remove set buffer py3

chpedersen committed:
CORE-4190: buffer_with_offset cython

chpedersen committed:
CORE-4190: simplify if

chpedersen committed:
CORE-4190: optimize udtype

chpedersen committed:
CORE-4190: optimize udtype

chpedersen committed:
CORE-4190: remove cython compiled check

chpedersen committed:
CORE-4190: remove not cython compiled case

chpedersen committed:
CORE-4190: remove not cython compiled case

chpedersen committed:
CORE-4190: remove not cython compiled case

chpedersen committed:
CORE-4190: remove not cython compiled case

chpedersen committed:
CORE-4190: remove not cython compiled case

chpedersen committed:
CORE-4190: remove not cython compiled case

chpedersen committed:
CORE-4190: remove not cython compiled case

chpedersen committed:
CORE-4190: remove not cython compiled case

chpedersen committed:
CORE-4190: proxied_dict py3 iterators

chpedersen committed:
CORE-4190: don't import range

chpedersen committed:
CORE-4190: add six & clru requirements

chpedersen commented:
> I'm getting a failure in python 3.9:
> 
> ```
> ======================================================================
> FAIL: testProxiedTupleHash (tests.mapped_struct.ProxiedTuplePackingTest)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/home/claudiofreire/src/sharedbuffers/tests/mapped_struct.py", line 2075, in testProxiedTupleHash
>     self.assertTrue(hash(p) == hash(t))
> AssertionError: False is not true
> 
> ----------------------------------------------------------------------
> Ran 637 tests in 1901.311s
> ```
> 
> I suspect it has something to do with python using randomized hash seeds.

The hash function for tuples was changed: https://github.com/python/cpython/pull/9471/files so I guess we have to translate the new version to python and use that one for py 3.8+. BTW why are we doing this (in `proxied_tuple.__hash__`), instead of just `self._hash = hash(tuple(iter(self)))` like below?

chpedersen commented:
Since we are now requiring cython, I removed the `not cython.compiled` cases that I added _in this PR_

chpedersen committed:
CORE-4190: Python 3.8 tuple hash in proxied_tuple

chpedersen committed:
CORE-4190: Python 3.8 version global

klaussfreire commented:
> > I'm getting a failure in python 3.9:
> > ```
> > ======================================================================
> > FAIL: testProxiedTupleHash (tests.mapped_struct.ProxiedTuplePackingTest)
> > ----------------------------------------------------------------------
> > Traceback (most recent call last):
> >   File "/home/claudiofreire/src/sharedbuffers/tests/mapped_struct.py", line 2075, in testProxiedTupleHash
> >     self.assertTrue(hash(p) == hash(t))
> > AssertionError: False is not true
> > 
> > ----------------------------------------------------------------------
> > Ran 637 tests in 1901.311s
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I suspect it has something to do with python using randomized hash seeds.
> 
> The hash function for tuples was changed: https://github.com/python/cpython/pull/9471/files so I guess we have to translate the new version to python and use that one for py 3.8+. BTW why are we doing this (in `proxied_tuple.__hash__`), instead of just `self._hash = hash(tuple(iter(self)))` like below?

Because it's faster. Creating a tuple only to hash it has move overhead.

klaussfreire commented:
> The hash function for tuples was changed: https://github.com/python/cpython/pull/9471/files so I guess we have to translate the new version to python and use that one for py 3.8+. BTW why are we doing this (in `proxied_tuple.__hash__`), instead of just `self._hash = hash(tuple(iter(self)))` like below?

Beware, there are 2 contexts in which we may require a hash.

To implement `__hash__`, in which case we can do whatever (keep the old or use the new), each option has its pros and cons.

And to use as hashing for the object mapper (stablehash), in which case we **cannot** update it, as it would break binary compatibility.

In general, there's no need to update the algorithm. We can keep the old. In practice, the change in py3 was made for a reason, and we might want to follow suit, but not at the cost of breaking binary compatibility. So I would not change stablehash.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -135,35 +135,35 @@
     if size is not None:
         if python3:
-            return buffer(data)[offset:offset+size]
+            return memoryview(data).cast("B")[offset:offset+size]
```
You should probably call make_memoryiew to centralize the logic there

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -135,35 +135,35 @@
 def make_memoryview(a):
-    return memoryview(a).cast("B")
+    if python3:
+        return memoryview(a).cast("B")
```
Can we avoid creating the memoryview?

``` python
if type(a) is not memoryview:
    a = memoryview(a)
```

There's probably a way to also avoid the cast... it should be a fast way, `a.format == "B"` works but it may be slow if it's python operations instead of native. I could not find a way to make it native.

I did a test, and it seems even without cython help it's quite faster to avoid the unnecessary operations:

``` python
import cython
  
_memoryview = cython.declare(object)
_memoryview = memoryview

def fbuffer(a):
    rv = a if type(a) is _memoryview else _memoryview(a)
    if rv.format != 'B':
        rv = rv.cast('B')
    return rv

def mbuffer(a):
    rv = a if type(a) is _memoryview else _memoryview(a)
    return rv.cast('B')

def sbuffer(a):
    return _memoryview(a).cast('B')
```

With a few timeits:

```
>>> import timeit, cython, pyximport
>>> pyximport.install(pyimport=True)
(<pyximport.pyximport.PyImporter object at 0x7ff42a518cc0>, <pyximport.pyximport.PyxImporter object at 0x7ff42a51ec18>)
>>> import ctest
>>> timeit.timeit(lambda:ctest.fbuffer(b"hola"), number=10000000) / 10.0
0.3661114629008807
>>> timeit.timeit(lambda:ctest.mbuffer(b"hola"), number=10000000) / 10.0
0.5427464027947281
>>> timeit.timeit(lambda:ctest.sbuffer(b"hola"), number=10000000) / 10.0
0.5551418196002487
>>> timeit.timeit(lambda:ctest.fbuffer(ctest.fbuffer(b"hola")), number=10000000) / 10.0
0.5739469211956021
>>> timeit.timeit(lambda:ctest.mbuffer(ctest.mbuffer(b"hola")), number=10000000) / 10.0
0.8258275890024379
>>> timeit.timeit(lambda:ctest.sbuffer(ctest.sbuffer(b"hola")), number=10000000) / 10.0
0.9486311056010891
```

It's clear that fbuffer wins by a healthy margin. It's a small difference (a fraction of a microsecond), but it's used a lot this make_memoryview thing, so I think it's worth paying attention to performance there.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -1309,19 +1301,28 @@ def iteritems(self):
         return iter(self.vlist)
 
     def values(self):
-        return self.vlist
```
Why not leave it as is? The vlist is iterable, so it's a valid return value.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -1309,19 +1301,28 @@ def iteritems(self):
     def keys(self):
-        return self.objmapper.keys()
+        if python3:
+            return iter(self.objmapper)
```
Why not just 

```suggestion
            return self.objmapper
```

It's iterable, so it's a valid return value (py3 returns an iterable set-like object not a pure iterator)

We're missing the set-iike part but maybe we can fix it in objmapper in some future PR.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -6595,13 +6546,10 @@ def build(cls, initializer, destfile = None, tempdir = None,
             else:
-                w = buffer(index)
+                w = make_memoryview(index)
             write(w)
```
What about

``` python
nitems = len(index)
if nitems != 0:
    write(make_memoryview(index))
```

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -7232,11 +7180,11 @@ def build(cls, initializer, destfile = None, tempdir = None, return_mapper = Tru
         else:
-            d = buffer(index)
+            d = make_memoryview(index)
         write(d)
```
Similarly here, I think it can be simplified.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -4394,7 +4628,7 @@ def get_packer(self, obj):
             ]))
```
I'm getting these errors in tests:

```
======================================================================
ERROR: testUnpack (tests.pool.SmallIntContainerPackingTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/claudiofreire/src/sharedbuffers/tests/pool.py", line 114, in testUnpack
    pos = p.pack(self.schema, x)[0]
  File "/home/claudiofreire/src/sharedbuffers/sharedbuffers/pool.py", line 225, in pack
    pos = self._pack_into(schema, obj, section)
  File "/home/claudiofreire/src/sharedbuffers/sharedbuffers/pool.py", line 171, in _pack_into
    buf = schema.pack(obj, idmap, implicit_offs=implicit_offs)
  File "sharedbuffers/mapped_struct.py", line 4809, in sharedbuffers.mapped_struct.Schema.pack
    def pack(self, obj, idmap = None, packer = None, padding = None, implicit_offs = 0):
  File "sharedbuffers/mapped_struct.py", line 4819, in sharedbuffers.mapped_struct.Schema.pack
    endp = self.pack_into(obj, buf, 0, idmap, packer, padding, implicit_offs)
  File "sharedbuffers/mapped_struct.py", line 4774, in sharedbuffers.mapped_struct.Schema.pack_into
    packer, padding = self.get_packer(obj)
  File "sharedbuffers/mapped_struct.py", line 4622, in sharedbuffers.mapped_struct.Schema.get_packer
    packer = struct.Struct("".join([
TypeError: sequence item 0: expected str instance, bytes found
```

Seems Struct.format is bytes, and the join there is using strings. Changing it to `b"".join` should fix it. There's quite a few places like these.

Edit: seems to affect python 3.6, not 3.9. Don't know about the in-between versions.

klaussfreire replied:
Yep, seems 3.6 has bytes in Struct.format, but not the other versions. Just doing b"".join doesn't fix it for all, it breaks in 3.7+.

Checking if there's a way that's worth the trouble. Otherwise we may have to drop support for 3.6-

klaussfreire replied:
Ok, it looks ugly, but this seems to work, and it's a small effort to get 3.6 support as it's only in one place, so might as well do it:

``` diff
diff --git a/sharedbuffers/mapped_struct.py b/sharedbuffers/mapped_struct.py
index 41f19ce..51f4df5 100644
--- a/sharedbuffers/mapped_struct.py
+++ b/sharedbuffers/mapped_struct.py
@@ -4618,9 +4618,12 @@ class Schema(object):
             packer_key = present_bitmap
 
         rv = self.packer_cache.get(packer_key)
+        bitmap_packer_format = self.bitmap_packer.format
+        if isinstance(bitmap_packer_format, bytes):
+            bitmap_packer_format = bitmap_packer_format.decode('ascii')
         if rv is None:
             packer = struct.Struct("".join([
-                self.bitmap_packer.format,
+                bitmap_packer_format,
             ] + [
                 self.slot_struct_types[slot]
                 for i,slot in enumerate(self.slot_keys)
```

chpedersen replied:
Yep, I was just doing something like this

chpedersen committed:
CORE-4190: struct format fix for py3.6

chpedersen committed:
CORE-4190: call make_memoryview

chpedersen committed:
CORE-4190: simplify write

chpedersen committed:
CORE-4190: simplify write

chpedersen committed:
CORE-4190: simplify write

klaussfreire committed:
CORE-4190: Fix memoryview refhandling

chpedersen committed:
CORE-4190: proxied_dict.keys

chpedersen committed:
CORE-4190: proxied_dict.values

chpedersen commented:
@klaussfreire

klaussfreire commented on tests/mapped_struct.py:
```diff
@@ -110,8 +117,8 @@ class PrimitiveStruct(TestStruct):
-        's' : str,
-        'u' : unicode,
+        's' : bytes,
+        'u' : six.text_type,
```
I prefer importing text_type as unicode, ends up bytes/unicode which is clear and unambiguous.

chpedersen replied:
If I do that, `udtype='unicode'` in `mapped_tuple.pack_into` and `mapped_frozenset.pack_into` doesn't work: "Not a type". I couldn't find a workaround.

klaussfreire replied:
I'm talking only about doing that in tests. There's no cython in tests.

klaussfreire commented on tests/mapped_struct.py:
```diff
@@ -2202,29 +2256,30 @@ def testHashFloatsFixed(self):
-            (frozenset([1.2, 4.5, 0.12]), 4015877951310865576L),
+            ((1, "abc", -2.3), long(2059039662577021167)),
+            # FIXME this doesnt work, the frozensets have differente order in py3 vs py2
+            # (frozenset([1.2, 4.5, 0.12]), 4015877951310865576),
```
This is a problem. This test is testing stable_hash, and as the name implies, it should be stable and not order dependent.

chpedersen replied:
IIUC, if the `_mix_hash` function were something like `code1 ^ code2` it would work, right? (I mean it wouldn't be order dependent, the hash would be different ofc)

```python
def _mix_hash(code1, code2):
    return _hash_rotl(code1, 5) ^ code2
def _stable_hash(key):
...
    elif isinstance(key, (tuple, frozenset, proxied_tuple, proxied_frozenset)):
        if isinstance(key, (frozenset, proxied_frozenset)):
            hval = _FSET_SEED
        else:
            hval = _TUPLE_SEED

        for value in key:
            hval = _mix_hash(hval, _stable_hash(value))
```

klaussfreire replied:
`code1 ^ code2` is just a bad mix function.

Imagine if `l` is some list, any list, `hash(l + l)` is always 0.

I suspect stable_hash is broken for python frozensets... in both versions... huh...

klaussfreire replied:
I'll look into it and maybe make a PR...

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -81,13 +84,19 @@
 import ctypes
 from datetime import timedelta, datetime, date
 from decimal import Decimal
+from six import itervalues, reraise, iteritems, iterkeys
```
As I mentioned somewhere else, you shouldn't use them in cythonized code. Cython does not recognize them and thus doesn't optimize them. We want optimized iteration in most cases.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -114,10 +128,23 @@
 
+if python3:
+    long = int
+    basestring = (bytes, str)
```
``` python
from past.builtins import basestring
```

But, again, these break cython's ability to perform optimizations.

I think it's best to trust cython. Cython understands unicode and bytes in both versions properly without any magic compatibility declarations.

I could only get accelerated isinstance checks by using `isinstance(x, (bytes, unicode))` literally, without any helper constant. You may have to do it that way when necessary.

Cython also understands basestring, but in py3 it becomes just str (unicode), does not include bytes. In some contexts, it may be correct, so pay attention to context.

klaussfreire replied:
Ping

klaussfreire reviewed (changes requested):
Check those six/past uses in mapped_struct, they inhibt optimizations.

In the meanwhile I'll give it a thorough inspection with cython.annotate to look for other issues.

klaussfreire committed:
Merge pull request #96 from jampp/CORE-4190-memoryview_refhandling

CORE-4190: Fix memoryview refhandling

chpedersen committed:
CORE-4190: use itervalues method

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -81,13 +84,19 @@
+try:
+    from chorde.clients.inproc import Cache
+except ImportError:
+    from clru.lrucache import LRUCache as Cache
```
Should not attempt to import chorde at all. Otherwise it still breaks with SyntaxError if installed from the py2 branch.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -2395,7 +2609,15 @@ def __repr__(self):
+
+XXPRIME_1 = cython.declare(cython.long, 11400714785074694791)
+XXPRIME_2 = cython.declare(cython.long, 14029467366897019727)
+XXPRIME_5 = cython.declare(cython.long, 2870177450012600261)
```
I think those are longlong

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -2412,24 +2634,49 @@ def __init__(self, *args, **kwargs):
+                    for i in range(len_):
+                        lane = hash(self[i])
+                        if lane == -1:
+                            return -1
```
IIUC that should not be necessary, as -1 is an error return code. Cython would already be checking it. Maybe double-check.

chpedersen replied:
yep
```c
+2653:                         lane = hash(self[i])
          __pyx_t_6 = __Pyx_GetItemInt(((PyObject *)__pyx_v_self), __pyx_v_i, Py_ssize_t, 1, PyInt_FromSsize_t, 0, 1, 1); if (unlikely(!__pyx_t_6)) __PYX_ERR(0, 2653, __pyx_L1_error)
          __Pyx_GOTREF(__pyx_t_6);
          __pyx_t_7 = PyObject_Hash(__pyx_t_6); if (unlikely(__pyx_t_7 == ((Py_hash_t)-1))) __PYX_ERR(0, 2653, __pyx_L1_error)
          __Pyx_DECREF(__pyx_t_6); __pyx_t_6 = 0;
          __pyx_v_lane = __pyx_t_7;
```

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -2412,24 +2634,49 @@ def __init__(self, *args, **kwargs):
+                    acc += len_ ^ (XXPRIME_5 ^ 3527539)
+                    if acc == -1:
+                        self._hash = 1546275796
+                        return self._hash
```
Don't return directly, let it fall back to the regular code path:

``` python
if acc == -1:
    acc = 1546275796
self._hash = acc
```

klaussfreire committed:
CORE-4190: Fix file leak in object pools

chpedersen committed:
CORE-4190: use iteritems method

klaussfreire committed:
CORE-4190: Add missing cleanup

chpedersen committed:
Merge remote-tracking branch 'origin/CORE-4190-migrate-py3' into CORE-4190-migrate-py3

chpedersen committed:
CORE-4190: use iterkeys method

chpedersen committed:
CORE-4190: remove extra import

klaussfreire committed:
Merge pull request #97 from jampp/CORE-4190-fix_file_leak

CORE-4190: Fix file leak in object pools

chpedersen committed:
CORE-4190: xxprime type

chpedersen committed:
CORE-4190: remove hash error check

chpedersen committed:
CORE-4190: fix hash return

klaussfreire committed:
CORE-4190: Py3 perf fixes

The many differences between Py3 and Py2 forced a few changes
that derailed performance, at times considerably. These are a few
fixes that recover most of the performance lost:

- cimport array.array to get direct access to the buffer
  for fast index iteration, as regular iteration is quite slow
  in py3
- Precompute bytes([x]) for all x at import time, and use a
  lookup list instead, as constructing the bytes manually, while
  relatively fast, is still slow compared to the py2 idiom it
  replaces and in a critical path
- Use mostly byte typecodes, avoid unicode as it forces an
  encode/decode step that's quite onerous. Still accept them
  from client code, but use bytes internally as much as possible.
- array(dtype, initializer) is very slow in py3. Replace it
  with the equivalent but somehow faster frombytes() when the
  initializer is a memoryview
- Avoid constructing arrays if a memoryview can be used instead,
  as py3 typed memoryviews behave like arrays in any case.
- Use array/memoryview tolist when possible, as it's quite optimized
  and much faster than list(x).
- Build idmapper indexes directly into numpy arrays. Py3 ints are
  slower than py2 ints by virtue of being arbitrary precision. The
  difference is tiny but it adds up. Building the index directly
  into a numpy array with cython's typed memoryviews support for
  direct and native access makes a big difference. Allows building
  big 1M arrays directly instead of using 1k batches, avoiding
  a whole sort step.

klaussfreire committed:
CORE-4190: use fromstring in py2

chpedersen committed:
CORE-4190: Use clru caches

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -2614,9 +2607,9 @@ def __repr__(self):
-XXPRIME_1 = cython.declare(cython.long, 11400714785074694791)
-XXPRIME_2 = cython.declare(cython.long, 14029467366897019727)
-XXPRIME_5 = cython.declare(cython.long, 2870177450012600261)
+XXPRIME_1 = cython.declare(cython.size_t, 11400714785074694791)
```
Py_hash_t if it's for python hashes, IIRC.

chpedersen replied:
`Py_uhash_t`

https://github.com/python/cpython/blob/08deed1af56bec8668c6cb4d5cfd89e393e1fe5e/Objects/tupleobject.c#L308

but I can't use it here, I tried `Py_uhash_t`, `cython.Py_uhash_t`, `"Py_uhash_t"` and it doesn't work (unknown type).

klaussfreire replied:
Yes, it seems Cython uses `Py_hash_t` internally in its pxd library, but doesn't export it to be used in user code. Weird.

Sounds like a bug if you ask me.

klaussfreire replied:
Leave it as `Py_ssize_t`, then, which is what `Py_hash_t` actually is after all. It's the same thing, but pyport.h says:

``` C
typedef Py_ssize_t Py_hash_t;
```

and you **can** use Py_ssize_t in cython code

chpedersen replied:
But it's unsigned, `uhash` not `hash`
```c
typedef size_t Py_uhash_t;
```

klaussfreire replied:
The constants are unsigned, but the return type of PyObject_Hash is signed. Whether you use unsigned or signed for intermediates depends on the calculations done, python's tuple code seems to do everything in unsigned and then cast to signed.

`\_(0_0)_/`

klaussfreire replied:
FTR, the original `long` was also signed. Maybe you meant `unsigned long` or, if you want to make sure it's 64 bit, `unsigned long long`.

chpedersen committed:
CORE-4190: XXPRIME types

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -2607,9 +2607,9 @@ def __repr__(self):
-XXPRIME_1 = cython.declare(cython.size_t, 11400714785074694791)
-XXPRIME_2 = cython.declare(cython.size_t, 14029467366897019727)
-XXPRIME_5 = cython.declare(cython.size_t, 2870177450012600261)
+XXPRIME_1 = cython.declare("unsigned long long", 11400714785074694791)
```
NB, but FTR there's `cython.ulonglong`

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -2412,24 +2626,46 @@ def __init__(self, *args, **kwargs):
+                        acc = (acc << 31) | (acc >> 33)
+                        acc *= XXPRIME_1
+                    acc += len_ ^ (XXPRIME_5 ^ 3527539)
+                    if acc == -1:
```
acc is size_t, which is unsigned, but you're comparing to a signed type. To get well defined behavior, you have to cast the literal:

```suggestion
                    if acc == cython.cast(cython.size_t, -1):
```

chpedersen committed:
CORE-4190: cast

chpedersen committed:
CORE-4190: ulonglong

klaussfreire committed:
Merge branch 'CORE-4190-migrate-py3' of github.com:jampp/sharedbuffers into CORE-4190-perf_fixes

klaussfreire committed:
CORE-4190: Buffer size constant

klaussfreire committed:
Merge pull request #98 from jampp/CORE-4190-perf_fixes

CORE-4190: Py3 perf fixes

klaussfreire committed:
CORE-4190: Fix basestring usage

basestring is not th same in py2 and py3, in py3 it only resolves
to str, and in py2 it's both unicode and bytes.

Cython understands "unicode" and "bytes" specially, generating
optimized type checks. Replacing basestring with (bytes, unicode)
generates better code.

No constant achieves the same result, sadly.

klaussfreire committed:
Merge pull request #99 from jampp/CORE-4190-basestring_fixes

CORE-4190: Fix basestring usage

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -1425,21 +1576,52 @@ def proxied_list_richcmp(a, b, op):
-        elif selfe > othere:
-            return 1
+        if isinstance(selfe, dict):
+            r = dict_cmp(selfe, othere)
+            if r:
+                return r
```
Why this?

If it's to support comparing lists that contain dicts, py3 doesn't support it, so it's fine if our proxied list doesn't either:

``` pycon
>>> [{3:4}] < [{}]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: '<' not supported between instances of 'dict' and 'dict'
```

chpedersen replied:
Ok... but this breaks `tests.mapped_struct.DictContainerPackingTest`
(also `None` vs `None` works in py2 but not in py3, there's a test that uses this)

chpedersen replied:
Maybe we can skip those tests for py3 (how?)

klaussfreire replied:
> Ok... but this breaks `tests.mapped_struct.DictContainerPackingTest`

Well, the test should not be testing something that's illegal in py3.

>  (also `None` vs `None` works in py2 but not in py3, there's a test that uses this)

Not sure what you mean by that. I can compare None-containing sutff quite well in py3. Do you have a case at hand?

> Maybe we can skip those tests for py3 (how?)

We could. At least those that should not be passing in py3, but lets check that it is actually the case before we start skipping tests.

chpedersen replied:
> Well, the test should not be testing something that's illegal in py3.

See `testPackUnpack` in `DictContainerPackingTest`, it tests that the packed->unpacked `pt`/`pl` are equal to the original values, and they are lists(or tuple) of dicts.

----

In `ProxiedListPackingTest.testProxiedListSpecificEqual`:
```python
        self.assertEquals(self.pack([1, None, 2.0]), (1, None, 2.0))
```

This uses `proxied_list_cmp`
```python
        if selfe < othere:
            return -1
        elif selfe > othere:
            return 1
```

and `selfe` and `othere` are both `None`. In py3 `TypeError: '<' not supported between instances of 'NoneType' and 'NoneType'`

So this is fixed with what I added:
```python
             if selfe is None and othere is None:
                 continue
```

klaussfreire replied:
That case is not about the dict. It's about an element being None directly. It can be implemented easily by just special-casing "selfe/othere is None"

chpedersen replied:
Yep, that's what I did. Sooo, what about the dict stuff?

klaussfreire replied:
I see the problem.

`proxied_list_richcmp` uses `proxied_list_cmp` to do the comparison, but dicts and frozensets don't have a total order in py3, and it raises an exception rather than invent a total order the way py2 does.

I mantain that we shouldn't be implementing comparisons that aren't valid in py3, so I think the best option is to implement a `proxied_list_eq` specializing in equality comparisons, to avoid depending on the inequalities for tuple values.

Then `proxied_list_richcmp` can use the eq one for `==` and `!=` checks, and `proxied_list_cmp` for the rest.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -1425,21 +1576,52 @@ def proxied_list_richcmp(a, b, op):
+            if r:
+                return r
+        else:
+            if selfe is None and othere is None:
```
```suggestion
            if selfe is othere:
```

It's a bit more general, and if they're the same object they're equal for sure. Might even help performance.

chpedersen committed:
CORE-4190: selfe is othere

chpedersen committed:
CORE-4190: proxied_list_eq

chpedersen commented on sharedbuffers/mapped_struct.py:
```diff
@@ -1560,21 +1560,23 @@ def proxied_list_richcmp(a, b, op):
-    _cmp = proxied_list_cmp(a, b)
     if op == 2: # Py_EQ
-        return _cmp == 0
+        return proxied_list_eq(a, b)
```
What's the difference between this and using `proxied_list_cmp` for everything? Is it that we don't use our custom `dict_cmp` when it's not necessary?

klaussfreire replied:
The difference is that we don't need a `dict_cmp` at all and can throw the same exception python dicts throw.

klaussfreire commented on sharedbuffers/mapped_struct.py:
```diff
@@ -1432,14 +1621,21 @@ def proxied_list_cmp(a, b):
-        elif selfe > othere:
-            return 1
+        if isinstance(selfe, dict):
+            r = dict_cmp(selfe, othere)
```
You can now remove dict_cmp and go back to the original implementation. Maybe keep the `selfe is othere` as an optimization.

chpedersen committed:
CORE-4190: don't use dict_cmp

chpedersen committed:
CORE-4190: rm dict_cmp

chpedersen committed:
CORE-4190: fix proxied_list __eq__

chpedersen committed:
CORE-4190: rm dict_cmp_py3

chpedersen committed:
CORE-4190: update VERSION & CHANGELOG

chpedersen committed:
Merge branch 'master' into CORE-4190-migrate-py3

# Conflicts:
#	doc/source/CHANGELOG.rst
#	setup.py
#	sharedbuffers/mapped_struct.py

klaussfreire commented on doc/source/CHANGELOG.rst:
```diff
@@ -5,6 +5,14 @@ All notable changes to this project will be documented here.
 
 The format is largely inspired by keepachangelog_.
 
+v1.0.0 - 2022-06-08
```
```suggestion
v1.0.0 - Unreleased
```

chpedersen committed:
Update doc/source/CHANGELOG.rst

Co-authored-by: klaussfreire <klaussfreire@gmail.com>

a-villagra commented on sharedbuffers/mapped_struct.py:
```diff
@@ -2216,13 +2422,14 @@ def _frozenset_eq(self, x):
     def __eq__(self, seq):
         if isinstance(self, proxied_frozenset):
+            return self._frozenset_eq(seq)
             return cython.cast(proxied_frozenset, self)._frozenset_eq(seq)
```
🤔

a-villagra commented on sharedbuffers/mapped_struct.py:
```diff
@@ -2352,12 +2560,14 @@ def _subset(self, seq, strict_subset):
     def __le__(self, seq):
         if isinstance(self, proxied_frozenset):
+            return self._subset(seq, False)
             return cython.cast(proxied_frozenset, self)._subset(seq, False)
```
yeah 🤔
If this is intended, delete these
